src/autolinks/autolinks.ts

```diff
@@ -210,15 +209,136 @@ export class Autolinks implements Disposable {
 				refsets.push([remote.provider, autoLinks]);
 			}
 		}
+	}
 
-		// Custom-configured autolinks
-		if (this._references.length && (remote?.provider == null || !options?.excludeCustom)) {
+	/** put custom-configured autolinks to mutable refsets */
+	private collectCustomAutolinks(remote: GitRemote | undefined, refsets: RefSet[]) {
+		if (this._references.length && remote?.provider == null) {
 			refsets.push([undefined, this._references]);
 		}
+	}
+
+	/**
+	 * it should always return non-0 result that means a probability of the autolink `b` is more relevant of the autolink `a`
+	 */
+	private static compareAutolinks(a: ComparingAutolinkSet, b: ComparingAutolinkSet) {
+		// consider that if the number is in the start, it's the most relevant link
+		if (b.index === 0) {
+			return 1;
+		}
+		if (a.index === 0) {
+			return -1;
+		}
+		// maybe it worths to use some weight function instead.
+		return (
+			b.autolink.prefix.length - a.autolink.prefix.length ||
+			-(b.startIndex - a.startIndex) ||
+			-(b.index - a.index)
+		);
+	}
+
+	/**
+	 * returns sorted list of autolinks. the first is matched as the most relevant
+	 */
+	async getBranchAutolinks(
+		branchName: string,
+		remote?: GitRemote,
+		options?: { excludeCustom: boolean },
+	): Promise<undefined | Autolink[]> {
+		const refsets: RefSet[] = [];
+		await this.collectIntegrationAutolinks(refsets);
+		await this.collectRemoteAutolinks(remote, refsets);
+		if (!options?.excludeCustom) {
+			this.collectCustomAutolinks(remote, refsets);
+		}
+		if (refsets.length === 0) return undefined;
+
+		return Autolinks._getBranchAutolinks(branchName, refsets);
+	}
+
+	static _getBranchAutolinks(branchName: string, refsets: Readonly<RefSet[]>) {
+		const autolinks = new Map<string, ComparingAutolinkSet>();
+
+		let match;
+		let num;
+		for (const [provider, refs] of refsets) {
+			for (const ref of refs) {
+				if (!isCacheable(ref)) {
+					continue;
+				}
+				if (ref.type === 'pullrequest' || (ref.referenceType && ref.referenceType !== 'branchName')) {
```

Comment:
Right now, there are no autolink references that match this condition: `(ref.referenceType && ref.referenceType !== 'branchName')`. We have not added `referenceType: 'commitMessage'` to any other autolink references. I would expect if we are doing this, that we should also update `_getAutoLinks` not to consider refs with `ref.referenceType === 'branchName'` or something equivalent.

However, I should caution: in one of my tests, I had all refs from my Jira integration that were not of `referenceType` `branchName`, but every reference, including those, still passed this check because they do not have a defined `referenceType`. I am glad that they did, because one of them matched the name of my branch: `GLVSC-593_check-expiration-paid-acc-offline`, and `getBranchAutolink` returned a match because of this.

I think, if we are treating cases where `referenceType` is not defined as "this ref applies to both branch names and commit messages", we are fine there - but I'm still curious if there are cases where we should be setting `referenceType: 'commitMessage'` anywhere, because those will help make this faster.

Reply:
The goal of the referenceType property is to cut off the case with no prefix from the commit message. 

>  in one of my tests, I had all refs from my Jira integration that were not of referenceType branchName, but every reference, including those, still passed this check because they do not have a defined referenceType

I expect this behavior, as all the refs from Jira are matched the expression `<prefix><id>` where the prefix is always not empty

>  I'm still curious if there are cases where we should be setting referenceType: 'commitMessage' anywhere, because those will help make this faster.

Now I don't know about these kinds of cases, but maybe in some future we will face with them. I used the referenceType property instead of simple binary flag because it's more easy to read and understand. If you prefer using flag like `onlyBranchNameRef`, let's do it